Initial ideas for an `AdminClient` wrapper

truemped opened the pull request:
This is an initial PR to get some early feedback on the approach. I've
tried to adapt to the existing style and code in the `client` namespace.

The `AdminClient` is fully async and returning instances of
`KafkaFuture`s. This PR contains a wrapper around this that allows
dereferencing the result (tested) and use `core.async` channels
(untested so far).

Refs #29

Thoughts? Best
Daniel

truemped committed:
Initial ideas for an `AdminClient` wrapper

This is an initial PR to get some early feedback on the approach. I've
tried to adapt to the existing style and code in the `client` namespace.

The `AdminClient` is fully async and returning instances of
`KafkaFuture`s. This PR contains a wrapper around this that allows
dereferencing the result (tested) and use `core.async` channels
(untested so far).

Refs #29

pyr commented on src/kinsky/admin.clj:
```diff
@@ -0,0 +1,87 @@
+  {:name      (.name listing)
+   :internal? (.isInternal listing)})
+
+(defn admin-client->driver
```
One thing which could be investigated here is to `extend-protocol` on `AdminClient` directly

truemped replied:
True! Will check, might be simpler

truemped replied:
I've used `extend-type` instead of `extend-protocol`. If I understand `extend-protocol`, then this is about providing implementations for one protocol in multiple types which is not the case here. So we can use `extend-type` directly.

pyr commented on src/kinsky/admin.clj:
```diff
@@ -0,0 +1,87 @@
+    (list-topics [this list-internal?]
+      (let [opts (doto (ListTopicsOptions.)
+                   (.listInternal list-internal?))]
+        (prn (.shouldListInternal opts))
```
leftover.

truemped replied:
Ups

pyr reviewed (commented):
I like the general direction.

truemped committed:
Use extend-type for the AdminClient directly

Instead of creating a wrapper object around the `AdminClient` we can
simply extend it with the `AdminClientDriver` protocol.

truemped committed:
Add topic creation

Rough edges for topic creation.

pyr commented:
I need a bit of time to review this, I should get to it soon, sorry for the wait.

truemped commented:
Absolutely no worries, I'm super slow right now. Currently you can only list and create topics, so it's not even close to being something like feature complete...

pyr reviewed (approved):
This seems to work as intended, I will merge it as is and it can be later improved from the main branch, thanks!
